### PR TITLE
feat: add `FeGaussianBlur` filter

### PR DESCRIPTION
**Pull request by jakex7**
# Summary

Continuation of #2316 
Introducing new filter `FeGaussianBlur`.

### Implementation notes

On Android there is no easy way to fully implement Gaussian blur, as there is no native api for this. While a basic implementation is possible with `RenderScript`, it does not allow for blur in one axis and greater than `25`

## Test Plan

Example app -> Filters -> FeGaussianBlur

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ✅     |
| Android |    ✅     |

### CODE REVIEW

File: apple/Filters/RNSVGFeGaussianBlur.mm [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+{
+  return concreteComponentDescriptorProvider<RNSVGFeGaussianBlurComponentDescriptor>();
+}
+
+- (void)updateProps:(Props::Shared const &)props oldProps:(Props::Shared const &)oldProps
+{
+  const auto &newProps = static_cast<const RNSVGFeGaussianBlurProps &>(*props);
+
+  self.in1 = RCTNSStringFromStringNilIfEmpty(newProps.in1);
+  self.stdDeviationX = [NSNumber numberWithFloat:newProps.stdDeviationX];
```
Comment (WoLewicki) [inline apple/Filters/RNSVGFeGaussianBlur.mm:38]: Is having both coordinates for stdDeviation part of the standard?

File: apple/Utils/RCTConvert+RNSVG.mm [hunk 2]
```diff
@@ -52,13 +52,13 @@ @implementation RCTConvert (RNSVG)
     intValue)
 
 RCT_ENUM_CONVERTER(
-    RNSVGEdgeModeTypes,
+    RNSVGEdgeMode,
```
Comment (WoLewicki) [inline apple/Utils/RCTConvert+RNSVG.mm:55]: Why did you rename it?

File: macos-example/macos/Podfile.lock [hunk 3]
```diff
@@ -1053,7 +1053,7 @@ PODS:
     - React-jsi (= 0.73.32)
     - React-logger (= 0.73.32)
     - React-perflogger (= 0.73.32)
-  - RNReanimated (3.9.0):
+  - RNReanimated (3.9.0-nightly-20240421-474f6983d):
```
Comment (WoLewicki) [inline macos-example/macos/Podfile.lock:1056]: Should it end up here?

File: src/filter-image/FilterImage.tsx [hunk 4]
```diff
@@ -13,6 +13,8 @@ const getFilters = (filters: FilterImageProps['filters']) => {
     switch (name) {
       case 'colorMatrix':
         return <FeColorMatrix key={`${filter}-${index}`} {...filterProps} />;
+      case 'gaussianBlur':
```
Comment (WoLewicki) [inline src/filter-image/FilterImage.tsx:16]: Those are not needed after the refactor yes?

File: src/lib/extract/extractFilter.ts [hunk 5]
```diff
@@ -53,3 +55,27 @@ export const extractFeColorMatrix = (
...
+    ) {
+      extracted.stdDeviationX = Number(props.stdDeviation) || 0;
+      extracted.stdDeviationY = Number(props.stdDeviation) || 0;
+    } else {
+      const stdDeviation = props.stdDeviation.split(spaceReg);
+      extracted.stdDeviationX = Number(stdDeviation[0]) || 0;
+      extracted.stdDeviationY = Number(stdDeviation[1]) || 0;
+    }
+  }
+  if (props.edgeMode) extracted.edgeMode = props.edgeMode;
```
Comment (WoLewicki) [inline src/lib/extract/extractFilter.ts:79]: use {}

File: src/lib/extract/extractFilter.ts [hunk 6]
```diff
@@ -53,3 +55,27 @@ export const extractFeColorMatrix = (
 
   return extracted;
 };
+
+export const extractFeGaussianBlur = (
+  props: FeGaussianBlurComponentProps
+): FeGaussianBlurNativeProps => {
+  const extracted: FeGaussianBlurNativeProps = {};
+
+  if (props.in) extracted.in1 = props.in;
```
Comment (WoLewicki) [inline src/lib/extract/extractFilter.ts:64]: use {}

File: src/lib/extract/extractFilter.ts [hunk 7]
```diff
@@ -53,3 +55,27 @@ export const extractFeColorMatrix = (
...
+  if (props.stdDeviation !== undefined) {
```
Comment (WoLewicki) [inline src/lib/extract/extractFilter.ts:65]: What if it is undefined?

File: src/lib/extract/extractFilter.ts [hunk 8]
```diff
@@ -53,3 +55,27 @@ export const extractFeColorMatrix = (
...
+    if (
```
Comment (WoLewicki) [inline src/lib/extract/extractFilter.ts:66]: Once again, is splitting those values part of the standard?

**Review by WoLewicki — COMMENTED** [review]
Left some comments. Please answer them!

File: apple/Filters/RNSVGFeGaussianBlur.mm (hunk 1 above)
Comment (jakex7) [reply apple/Filters/RNSVGFeGaussianBlur.mm:38]: Yes, according to the W3 specs, there is a `stdDeviationX` as well as `stdDeviationY` https://www.w3.org/TR/SVG11/filters.html#InterfaceSVGFEGaussianBlurElement

File: macos-example/macos/Podfile.lock (hunk 3 above)
Comment (jakex7) [reply macos-example/macos/Podfile.lock:1056]: Somehow, the `main` branch had wrong `Podfile.lock` and the CI were failing for this one without the update.

File: src/lib/extract/extractFilter.ts (hunk 8 above)
Comment (jakex7) [reply src/lib/extract/extractFilter.ts:66]: Yes, according to W3 specs https://www.w3.org/TR/SVG11/filters.html#feGaussianBlurStdDeviationAttribute

> stdDeviation = "[number-optional-number](https://www.w3.org/TR/SVG11/types.html#DataTypeNumberOptionalNumber)"

> The standard deviation for the blur operation. If two [number](https://www.w3.org/TR/SVG11/types.html#DataTypeNumber)s are provided, the first number represents a standard deviation value along the x-axis of the coordinate system established by attribute [‘primitiveUnits’](https://www.w3.org/TR/SVG11/filters.html#FilterElementPrimitiveUnitsAttribute) on the [‘filter’](https://www.w3.org/TR/SVG11/filters.html#FilterElement) element. The second value represents a standard deviation in Y. If one number is provided, then that value is used for both X and Y.
...
A value of zero disables the effect of the given filter primitive (i.e., the result is the filter input image). If [‘stdDeviation’](https://www.w3.org/TR/SVG11/filters.html#feGaussianBlurStdDeviationAttribute) is 0 in only one of X or Y, then the effect is that the blur is only applied in the direction that has a non-zero value.

File: src/lib/extract/extractFilter.ts (hunk 7 above)
Comment (jakex7) [reply src/lib/extract/extractFilter.ts:65]: Hmm actually this check may not be needed as `defaultProps` should be applied in that case (with `stdDeviation: 0`)

File: android/src/main/java/com/horcrux/svg/FeGaussianBlurView.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+  @Override
+  public Bitmap applyFilter(HashMap<String, Bitmap> resultsMap, Bitmap prevResult) {
+    Bitmap source = getSource(resultsMap, prevResult, this.mIn1);
+    return blur(getContext(), source);
+  }
+
+  private Bitmap blur(Context context, Bitmap bitmap) {
+    // Android blur radius is much weaker than SVG's, so we need to scale it up.
+    float stdDeviation = Math.max(mStdDeviationX, mStdDeviationY) * 2;
```
[the remark this answers is not in the document]
Comment (jakex7) [reply android/src/main/java/com/horcrux/svg/FeGaussianBlurView.java:52]: `radius` on android is a different thing than `stdDeviation` anyway, and based on my tests, 2 will give the closest results compared to the web.

File: apple/Utils/RCTConvert+RNSVG.mm (hunk 2 above)
Comment (jakex7) [reply apple/Utils/RCTConvert+RNSVG.mm:55]: I believe that it is more accurate

File: src/lib/extract/extractFilter.ts (hunk 7 above)
Comment (jakex7) [reply src/lib/extract/extractFilter.ts]: Removed in https://github.com/software-mansion/react-native-svg/pull/2352/commits/8ffa742efeb7dac6511943813a2b38e4a2fae0b7

File: src/lib/extract/extractFilter.ts (hunk 6 above)
Comment (jakex7) [reply src/lib/extract/extractFilter.ts]: https://github.com/software-mansion/react-native-svg/pull/2352/commits/8ffa742efeb7dac6511943813a2b38e4a2fae0b7

File: src/lib/extract/extractFilter.ts (hunk 5 above)
Comment (jakex7) [reply src/lib/extract/extractFilter.ts]: https://github.com/software-mansion/react-native-svg/pull/2352/commits/8ffa742efeb7dac6511943813a2b38e4a2fae0b7

File: android/src/main/java/com/horcrux/svg/FeGaussianBlurView.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    if (stdDeviation <= 0) return bitmap;
+    final float maxRadius = 25.0f;
+    float radius = Math.min(stdDeviation, maxRadius);
+
+    Bitmap outputBitmap = Bitmap.createBitmap(bitmap);
+
+    RenderScript rs = RenderScript.create(context);
+    ScriptIntrinsicBlur blurScript = ScriptIntrinsicBlur.create(rs, Element.U8_4(rs));
```
[the remark this answers is not in the document]
Comment (jakex7) [reply android/src/main/java/com/horcrux/svg/FeGaussianBlurView.java:61]: Added in https://github.com/software-mansion/react-native-svg/pull/2352/commits/8ffa742efeb7dac6511943813a2b38e4a2fae0b7

File: android/src/main/java/com/horcrux/svg/RenderableViewManager.java [hunk 11]
```diff
@@ -1387,6 +1392,61 @@ public void setValues(FeColorMatrixView node, @Nullable ReadableArray values) {
...
+      implements RNSVGFeGaussianBlurManagerInterface<FeGaussianBlurView> {
+    FeGaussianBlurManager() {
+      super(SVGClass.RNSVGFeGaussianBlur);
+      mDelegate = new RNSVGFeGaussianBlurManagerDelegate(this);
+    }
+
+    public static final String REACT_CLASS = "RNSVGFeGaussianBlur";
+
+    @ReactProp(name = "x")
+    public void setX(FeGaussianBlurView node, Dynamic x) {
```
[the remark this answers is not in the document]
Comment (jakex7) [reply android/src/main/java/com/horcrux/svg/RenderableViewManager.java:1405]: Sure, that's a good idea. I'll do it in the follow-up PR.

File: android/src/main/java/com/horcrux/svg/FeGaussianBlurView.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (WoLewicki) [reply android/src/main/java/com/horcrux/svg/FeGaussianBlurView.java:52]: Would be nice to have it described somewhere then. Now it looks like a magic variable.
